### PR TITLE
Documentation for Foundations of Data Science Scripts

### PR DESCRIPTION
**Pull request by sweanan**
Closes [138](https://github.com/microsoft/dstoolkit-mlops-v2/issues/138)

This Document provides the Foundational Structure for the Data Science scripts with respect to Model Factory Architecture.

How the Data Science scripts are structured can greatly ease the integration of the code into Model Factory, This document provides the best practices to start drafting the DS scripts

### CODE REVIEW

File: docs/how-to/DataScienceScripts.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
+# Data Science Scripts - Foundational Structure: How to start drafting the DS scripts with respect to Model Factory Architecture
+
+## Context and Scope
+
+Data scientists perform exploratory data analysis (EDA) to understand the training data, and then perform experiments to create machine learning code that tunes a base model for the hypothesis based on the analysis of the training data. Data scientists can perform this work on their own workstation, on a Data Science Virtual Machine (DSVM), or on an Azure Machine Learning Compute instance.
+
+Data Scientists need a flexible development environment with some engineering fundamentals such as main branch protection for PR review as well as a simple CI pipeline with python linting, secrets checking, markdown linting, etc. This environment let's the data science experiment with models running on Azure Machine Learning compute without impacting production code.
```
Comment (marshallbentley) [inline docs/how-to/DataScienceScripts.md:7]: "This environment let's the data science experiment"

should be

"This environment lets the data scientist experiment"

File: docs/how-to/DataScienceScripts.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        dataset_folder=pipeline_job_input,
+    )
+    score_model_cmp = gl_pipeline_components[2](
+        pred_folder=predict_model_cmp.outputs.pred_folder,
+    )
+    ```
+
+### Data Science Script Configuration
+
+The Data Science script must have configuration, which defines the configuration parameter required for the data science scripts. The configuration file should be in yaml format and should be easily configurable. For example, the configuration file for the nyc_taxi model will be stored in the `src/nyc_taxi/common/nyc_config.yaml` folder. And the DS scripts can read the configuration file and use the parameters defined in it.  Below is an example of how the `nyc_taxi_config.yaml` file looks like:
```
Comment (marshallbentley) [inline docs/how-to/DataScienceScripts.md:68]: "defines the configuration parameter required"

should be

"defines the configuration parameters required"

Comment (marshallbentley) [inline docs/how-to/DataScienceScripts.md:68]: "Below is an example of how the `nyc_taxi_config.yaml` file looks like"

should be

"Below is an example of the `nyc_taxi_config.yaml` file"

File: docs/how-to/DataScienceScripts.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    import argparse
+    import mlflow
+
+    args = parser.parse_args()
+    mlflow.log_params(vars(args))
+    ```
+
+### Data Science Script Testing
+
+The Data Science script must be tested thoroughly to ensure that it works as expected. The testing must cover all the components of the script and should be automated as part of the CI pipeline to catch errors prior to merging code into `main`.
```
Comment (marshallbentley) [inline docs/how-to/DataScienceScripts.md:128]: We would not want automated tests to run as part of CI in the prototyping repo.  If we run them, it should be in the model factory during promotion.  The only tests in the prototyping would be the inference tests which are promoted alongside the model.  These however, would not be run as part of CI.

File: docs/how-to/DataScienceScripts.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+The Data Science scripts must be run individually as a python script and tested manually. However, in future, the testing can be automated using the [PYTest](https://docs.pytest.org/en/stable/getting-started.html#get-started) framework. The test files should be placed in the `test` folder of the DSToolkit.
```
Comment (marshallbentley) [inline docs/how-to/DataScienceScripts.md:130]: This conflicts with the statement above about automated tests.

File: docs/how-to/DataScienceScripts.md (hunk 1 above)
Comment (sweanan) [reply docs/how-to/DataScienceScripts.md]: Thanks @marshallbentley fixed it

File: docs/how-to/DataScienceScripts.md (hunk 2 above)
Comment (sweanan) [reply docs/how-to/DataScienceScripts.md]: Thanks @marshallbentley fixed it

Comment (sweanan) [reply docs/how-to/DataScienceScripts.md]: Thanks @marshallbentley fixed it

File: docs/how-to/DataScienceScripts.md (hunk 3 above)
Comment (sweanan) [reply docs/how-to/DataScienceScripts.md]: Thanks @marshallbentley changed it

**Review by marshallbentley — APPROVED** [review]

**Review by lorrinferdinand — APPROVED** [review]

**Pull request merged**